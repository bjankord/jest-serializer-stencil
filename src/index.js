const mockDoc = require('@stencil/core/mock-doc/index.cjs');
const toDiffableHTML = require('diffable-html');

const print = (val) => {
  const serializedHTML = mockDoc.serializeNodeToHtml(val, {
    serializeShadowRoot: true,
    prettyHtml: true,
    outerHtml: true,
  });
  return toDiffableHTML(serializedHTML).trim();
};

const test = (val) => {
  if (val !== undefined && val !== null) {
    if (val instanceof HTMLElement || val instanceof mockDoc.MockNode) {
      return true;
    }

    if (typeof val === 'string') {
      console.log('val is a string');
      const trimmed = val.trim();
      return (
        trimmed.length > 2 &&
        trimmed[0] === "<" &&
        trimmed[trimmed.length - 1] === ">"
      );
    }
  }

  return false;
}

module.exports = {
  print,
  test,
};
