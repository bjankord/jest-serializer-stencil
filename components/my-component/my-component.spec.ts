import { newSpecPage } from '@stencil/core/testing';
import { MyComponent } from './my-component';
import { print as printFn, test as testFn } from '../../src/index';

describe('my-component', () => {
  expect.addSnapshotSerializer({
    print(val) {
      return printFn(val);
    },
    test(val) {
      return testFn(val);
    },
  });

  it('renders with no attributes', async () => {
    const page = await newSpecPage({
      components: [MyComponent],
      html: '<my-component></my-component>',
    });
    await page.waitForChanges();

    expect(page.root).toMatchSnapshot();
  });

  it('renders with 1 attribute', async () => {
    const page = await newSpecPage({
      components: [MyComponent],
      html: '<my-component id="foo"></my-component>',
    });
    await page.waitForChanges();

    expect(page.root).toMatchSnapshot();
  });

  it('renders with 2 attributes', async () => {
    const page = await newSpecPage({
      components: [MyComponent],
      html: '<my-component id="foo" class="bar"></my-component>',
    });
    await page.waitForChanges();

    expect(page.root).toMatchSnapshot();
  });

  it('renders with multiple attributes', async () => {
    const page = await newSpecPage({
      components: [MyComponent],
      html: '<my-component id="foo" first="First" middle="Middle" last="Last" data-attr="bar"></my-component>',
    });
    await page.waitForChanges();

    expect(page.root).toMatchSnapshot();
  });
});
