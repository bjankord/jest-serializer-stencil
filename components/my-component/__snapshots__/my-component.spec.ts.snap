// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`my-component renders with 1 attribute 1`] = `
<my-component id="foo">
  <mock:shadow-root>
    <div>
      Hello, World! I'm
    </div>
  </mock:shadow-root>
</my-component>
`;

exports[`my-component renders with 2 attributes 1`] = `
<my-component
  class="bar"
  id="foo"
>
  <mock:shadow-root>
    <div>
      Hello, World! I'm
    </div>
  </mock:shadow-root>
</my-component>
`;

exports[`my-component renders with multiple attributes 1`] = `
<my-component
  data-attr="bar"
  first="First"
  id="foo"
  last="Last"
  middle="Middle"
>
  <mock:shadow-root>
    <div>
      Hello, World! I'm First Middle Last
    </div>
  </mock:shadow-root>
</my-component>
`;

exports[`my-component renders with no attributes 1`] = `
<my-component>
  <mock:shadow-root>
    <div>
      Hello, World! I'm
    </div>
  </mock:shadow-root>
</my-component>
`;
